src/bigquery.ts

```diff
@@ -2114,22 +2161,183 @@ export class BigQuery extends Service {
         : {};
     const callback =
       typeof optionsOrCallback === 'function' ? optionsOrCallback : cb;
-    this.createQueryJob(query, (err, job, resp) => {
+
+    this.trace_('[query]', query, options);
+    const queryReq = this.buildQueryRequest_(query, options);
+    this.trace_('[query] queryReq', queryReq);
+    if (!queryReq) {
+      this.createQueryJob(query, (err, job, resp) => {
+        if (err) {
+          (callback as SimpleQueryRowsCallback)(err, null, resp);
+          return;
+        }
+        if (typeof query === 'object' && query.dryRun) {
+          (callback as SimpleQueryRowsCallback)(null, [], resp);
+          return;
+        }
+        // The Job is important for the `queryAsStream_` method, so a new query
+        // isn't created each time results are polled for.
+        options = extend({job}, queryOpts, options);
+        job!.getQueryResults(options, callback as QueryRowsCallback);
+      });
+      return;
+    }
+
+    this.runJobsQuery(queryReq, (err, job, res) => {
+      this.trace_('[runJobsQuery callback]: ', query, err, job, res);
       if (err) {
-        (callback as SimpleQueryRowsCallback)(err, null, resp);
+        (callback as SimpleQueryRowsCallback)(err, null, res);
         return;
       }
-      if (typeof query === 'object' && query.dryRun) {
-        (callback as SimpleQueryRowsCallback)(null, [], resp);
+
+      options = extend({job}, queryOpts, options);
+      if (res && res.jobComplete) {
+        let rows: any = [];
+        if (res.schema && res.rows) {
+          rows = BigQuery.mergeSchemaWithRows_(res.schema, res.rows, {
+            wrapIntegers: options.wrapIntegers!, // TODO: fix default value
+            parseJSON: options.parseJSON,
+          });
+        }
+        options.cachedRows = rows;
```

Comment:
Do we ever surface this augmented options back to the user?  If so, do we need to redact the rows?

Reply:
added a comment and `_` to the property to make it clear that is a internal property. Also this is not gonna leak to users, as we deep clone user passed options.